[NIFI-10754] Initial check in of new getUri NIFI Expression Language …

dan-s1 opened the pull request:
…function.

<!-- Licensed to the Apache Software Foundation (ASF) under one or more -->
<!-- contributor license agreements.  See the NOTICE file distributed with -->
<!-- this work for additional information regarding copyright ownership. -->
<!-- The ASF licenses this file to You under the Apache License, Version 2.0 -->
<!-- (the "License"); you may not use this file except in compliance with -->
<!-- the License.  You may obtain a copy of the License at -->
<!--     http://www.apache.org/licenses/LICENSE-2.0 -->
<!-- Unless required by applicable law or agreed to in writing, software -->
<!-- distributed under the License is distributed on an "AS IS" BASIS, -->
<!-- WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. -->
<!-- See the License for the specific language governing permissions and -->
<!-- limitations under the License. -->

# Summary

[NIFI-10754](https://issues.apache.org/jira/browse/NIFI-10754)

# Tracking

Please complete the following tracking steps prior to pull request creation.

### Issue Tracking

- [x] [Apache NiFi Jira](https://issues.apache.org/jira/browse/NIFI) issue created

### Pull Request Tracking

- [x] Pull Request title starts with Apache NiFi Jira issue number, such as `NIFI-00000`
- [x] Pull Request commit message starts with Apache NiFi Jira issue number, as such `NIFI-00000`

### Pull Request Formatting

- [x] Pull Request based on current revision of the `main` branch
- [x] Pull Request refers to a feature branch with one commit containing changes

# Verification

Please indicate the verification steps performed prior to pull request creation.

### Build

- [x] Build completed using `mvn clean install -P contrib-check`
  - [x] JDK 8
  - [ ] JDK 11
  - [ ] JDK 17

### Licensing

- [ ] New dependencies are compatible with the [Apache License 2.0](https://apache.org/licenses/LICENSE-2.0) according to the [License Policy](https://www.apache.org/legal/resolved.html)
- [ ] New dependencies are documented in applicable `LICENSE` and `NOTICE` files

### Documentation

- [x] Documentation formatting appears as expected in rendered files

dan-s1 commented:
@exceptionfactory If you have time, can you please review this in order for this to get into 1.19?

dan-s1 commented:
@exceptionfactory Now that 1.19.0 has been released, when you have some time can you please review this?

dan-s1 commented:
@exceptionfactory Wondering if this could be revisited. It is functionality my team could use. Thanks!

exceptionfactory commented on nifi-commons/nifi-expression-language/src/main/java/org/apache/nifi/attribute/expression/language/evaluation/functions/GetUriEvaluator.java:
```diff
@@ -0,0 +1,77 @@
+    public QueryResult<String> evaluate(EvaluationContext evaluationContext) {
+        List<String> args = Arrays.stream(uriArgs)
+                .map(uriArg -> uriArg.evaluate(evaluationContext).getValue())
+                .map(string -> StringUtils.isBlank(string) ? null : string)
```
The `StringUtils.isBlank()` check would convert both empty strings such as `""` as well as strings with spaces such as `"  "` to `null`. Although these seem to be unlikely arguments, it raises a larger question about passing empty arguments. The check should probably be simplified to check for an empty string:
```suggestion
                .map(string -> string == null || string.isEmpty() ? null : string)
```

exceptionfactory reviewed (commented):
Thanks for working on this new feature @dan-s1. Although most of the implementation makes sense, it raises some questions about treating blanks strings as nulls. Although the `java.net.URI` has a number of constructors, it seems like this may provide more options than necessary. There is some value in simplicity, versus exposing all the possible options, which could also avoid potential confusion related to handling blank strings as nulls.

The discussion on NIFI-10754 originated out of issues with the `urlEncode` function. If the implementation were to be scoped down to just something like `uriEncodePath`, for instance, would that provide the desired functionality?

dan-s1 commented:
> Thanks for working on this new feature @dan-s1. Although most of the implementation makes sense, it raises some questions about treating blanks strings as nulls. Although the `java.net.URI` has a number of constructors, it seems like this may provide more options than necessary. There is some value in simplicity, versus exposing all the possible options, which could also avoid potential confusion related to handling blank strings as nulls.
> 
> The discussion on NIFI-10754 originated out of issues with the `urlEncode` function. If the implementation were to be scoped down to just something like `uriEncodePath`, for instance, would that provide the desired functionality?

@exceptionfactory  If I understand you correctly that means there would have to be potentially two calls when constructing a URI, one to urlEncodePath and one to urlEncode to encode the parameters. I was actually intending to simplify that process of URI construction with one call which would take care of the encoding for both the path and the parameters.

exceptionfactory commented:
Thanks for clarifying the intent @dan-s1, constructing the complete URI in a single function makes sense.

With that background, do you need all of the possible constructor options? It might be simpler to support just one or two options, as opposed to all of the convenience constructors.

To the other question, is there a possibility that an empty string should be allowed as an argument instead of treated as `null`? That kind of implicit conversion to `null` seems like it could create confusion, and might actually be a reason to provide a different set of supported arguments so that particular values can be `null` when passed to the `URI` constructors.

dan-s1 commented:
> Thanks for clarifying the intent @dan-s1, constructing the complete URI in a single function makes sense.
> 
> With that background, do you need all of the possible constructor options? It might be simpler to support just one or two options, as opposed to all of the convenience constructors.
> 
That is a hard call because I am not sure what different people need.

> To the other question, is there a possibility that an empty string should be allowed as an argument instead of treated as `null`? That kind of implicit conversion to `null` seems like it could create confusion, and might actually be a reason to provide a different set of supported arguments so that particular values can be `null` when passed to the `URI` constructors.
Per the URI javadocs

> a component of the new URI may be left undefined by passing null for the corresponding parameter.
I did not see a way of explicitly passing null except if a blank string was passed. Please let me know otherwise if null can be passed.

exceptionfactory commented:
Thanks for the reply @dan-s1.

As you observed, the Expression Language parser does not support a literal `null` as a function argument. It would take a bit more work, but it seems worth evaluating what it would take to support a literal `null` argument for `getUri`. It seems like it could be scoped to this function.

For `getUri` itself, there does not seem to be much value in the three-argument non-hierarchical option for `mailto` and similar links. This could be added later, but I think it makes more sense to focus on the hierarchical URI constructors unless you have a specific use case in mind for non-hierarchical URIs.

For hierarchical URIs, the four and five argument constructors appear to fall into the convenience category. For simplicity of the initial implementation, what do you think of limiting the implementation to the seven argument constructor? Although it is the most verbose, it provides the clearest implementation for documentation and usage.

dan-s1 commented:
@exceptionfactory I am fine with only implementing the the seven argument constructor. What did you have in mind for implementing a literal null as a function argument? I only question if it is more clearer than specifying to users to use an empty string which we can convert to null on the back end.

exceptionfactory commented:
> What did you have in mind for implementing a literal null as a function argument? I only question if it is more clearer than specifying to users to use an empty string which we can convert to null on the back end.

I was thinking that `null` would be handled as a literal along the following lines:

```
${getUri('https', null, 'nifi.apache.org', 8443, '/docs.html', null, null)}
```

An empty string could be passed to any of the string arguments, but it allowing a `null` would avoid the need to convert the arguments and clarify expected behavior.

dan-s1 commented:
@exceptionfactory I added the null literal as you requested and modified the documentation to reflect this. Please let me know if my changes are what you were intending. Thanks!

exceptionfactory commented on nifi-commons/nifi-expression-language/src/main/java/org/apache/nifi/attribute/expression/language/evaluation/functions/GetUriEvaluator.java:
```diff
@@ -0,0 +1,63 @@
+
+        try {
+            if (args.size() == 7) {
+                return new StringQueryResult(new URI(args.get(0), args.get(1), args.get(2),
+                        Integer.parseInt(args.get(3)), args.get(4), args.get(5), args.get(6)).toString());
```
Recommend breaking this into multiple lines for readability. I also recommend moving the integer parsing to a separate private method, where the `NumberFormatException` could be caught apart from the others.
```suggestion
                final String scheme = args.get(0);
                final String userInfo = args.get(1);
                final String host = args.get(2);
                final int port = getPort(args.get(3));
                final String path = args.get(4);
                final String query = args.get(5);
                final String fragment = args.get(6);
                final URI uri = new URI(scheme, userInfo, host, port, path, query, fragment);
                return new StringQueryResult(uri.toString());
```

dan-s1 replied:
@exceptionfactory Done

exceptionfactory reviewed (changes requested):
Thanks for making the adjustments @dan-s1, this looks close to completion. I noted one more substantive change, breaking out the URI constructor arguments and port parsing to make the code more readable.

dan-s1 commented:
@exceptionfactory I noticed the build ci-workflow / Ubuntu Zulu JDK 11 HI failed but it failed on a unit test unrelated to my changes. Also the ci-workflow / Ubuntu Zulu JDK 17 EN timed out. Please advise.

dan-s1 commented:
@exceptionfactory Is there anything else that needs to be done for this PR?

exceptionfactory reviewed (approved):
Thanks for working through the feedback @dan-s1, the latest version looks good from a code and functional perspective. I noticed a minor documentation formatting issue with the first example expression usage, but I will correct that in the merge commit. +1 merging
